src/BlazorWebView/src/core/Android/BlazorWebViewHandler.Android.cs

```diff
@@ -54,13 +56,42 @@ protected override void DisconnectHandler(AWebView nativeView)
 			Services != null;
 
 		private string? HostPage { get; set; }
-		private ObservableCollection<RootComponent>? RootComponents { get; set; }
-		private new IServiceProvider? Services { get; set; }
+		private ObservableCollection<RootComponent>? RootComponents
+		{
+			get => _rootComponents;
+			set
+			{
+				if (_rootComponents != null)
+				{
+					// Remove any previously-known root components and unhook events
+					_rootComponents.Clear();
+					_rootComponents.CollectionChanged -= OnRootComponentsCollectionChanged;
+				}
+
+				_rootComponents = value;
+
+				if (_rootComponents != null)
+				{
+					// Add new root components and hook events
+					if (_rootComponents.Count > 0 && _webviewManager != null)
+					{
+						_ = _webviewManager.Dispatcher.InvokeAsync(async () =>
+						{
+							foreach (var component in _rootComponents)
+							{
+								await component.AddToWebViewManagerAsync(_webviewManager);
+							}
+						});
+					}
+					_rootComponents.CollectionChanged += OnRootComponentsCollectionChanged;
```

Comment:
There might be a race condition here. If L86 runs before the dispatcher continuation from L80, and the collection changes in between those two executions, then I think a root component would be added twice.

In practice it probably wouldn't happen because we're probably already on the dispatcher thread, but just in case, maybe it's worth *not* dispatching, and instead using `Dispatcher.CheckAccess()` (throwing if false) to be sure we didn't need to.

Reply:
I think I've fixed this now (but the code is in another file now). But if equality works the way we think, I think the logic in the CollectionChanged handler would avoid any potential problem anyway, no?